# Zero to Hero

## 1.06 Hackeando Kaggle

Voy a **hackear** Kaggle  para descubrir cuantos  "BAJA+2"  hay en el mes 202101, que es donde  *no* tengo la clase

La idea es subir a Kaggle una prediccion que tenga TODOS 1,  que se le envíe estímulo a todos
<br>
Es obvio que esa prediccion va a dar una pésima ganancia
<br>
Pero me va a permitir tener un sistema de dos ecuaciones con dos incognitas, y despejando tendre la cantidad de BAJA+2 que hay en los datos del futuro

In [1]:
rm( list=ls())
gc()

library( "data.table")   #cargo la libreria  data.table
library( "rpart")  #cargo la libreria  rpart
library( "rpart.plot")

used   (Mb) gc trigger (Mb) max used (Mb)
Ncells 474386 25.4 1044432    55.8  621911  33.3
Vcells 875003  6.7 8388608    64.0 1601572  12.3

Warning message:
"package 'data.table' was built under R version 3.6.3"Warning message:
"package 'rpart' was built under R version 3.6.3"

In [2]:
options(repr.plot.width=20, repr.plot.height=10) 
setwd("C:\\Users\\oliva\\OneDrive\\Desktop\\DS\\Austral\\08 - Labo1") #Aqui se debe poner la ruta de la PC local

Leo el dataset de **enero 2021**

In [8]:
denero <- fread("./labo/datasets/paquete_premium_202101.csv")   #cargo el dataset

cuento la cantidad de lineas del dataset

In [9]:
nrow( denero )

[1] 160607

Creo un vector que sean los  ids

In [12]:
vector_ids  <-   denero[ , numero_de_cliente] #todas las filas, solo la comuna numero de clientes

In [13]:
head( vector_ids)

[1] 4572300 4572957 4572973 4572998 4573029 4573199

Creo un vector de todos unos, de  160607 unos,  la cantidad de registros que tiene denero
<br>
utilizo la instruccion  **rep()**  de R  https://www.rdocumentation.org/packages/base/versions/3.6.2/topics/rep

In [14]:
vector_enviar <-  rep( 1,  nrow(denero))

In [15]:
head(vector_enviar)

[1] 1 1 1 1 1 1

In [16]:
length(vector_enviar)

[1] 160607

Excelente, tiene tantos unos como registros tiene denero

creo una tabla con  lo que voy a enviar a Kaggle

In [17]:
tabla_final  <-   as.data.table(  list(  "numero_de_cliente"= vector_ids,
                                         "Predicted"=         vector_enviar))

In [18]:
head( tabla_final)

numero_de_cliente Predicted
1 4572300           1        
2 4572957           1        
3 4572973           1        
4 4572998           1        
5 4573029           1        
6 4573199           1

In [22]:
#genero el archivo para Kaggle
#creo la carpeta donde va el experimento
dir.create( "./labo/exp/ZH2016/", showWarnings = FALSE )

#genero el archivo para Kaggle
fwrite(tabla_final, 
        file= "./labo/exp/ZH2016/todos_unos.csv",
        sep=  "," )

Subo el archivo a Kaggle y me fijo cuanta ganancia me da en el Public Leaderboard

Da una ganancia de **-115.307**

### El hackeo

Ahora pasamos a calcular la cantidad de BAJA+2  que hay en los datos de enero-2021,  hackeamos Kaggle

Tengo dos incognitas, por un lado lo que llamo POS , los positivos, las "BAJA+2"
<br>
Por otro lado, lo que llamo NEG, los negativos,  los "BAJA+1" y "CONTINUA"

160607Se que se cumplen estas dos cosas:
* POS + NEG = 160607
* 59000\*POS - 1000\*NEG =  -115307000

Es un sistema de dos ecuaciones con dos incógnitas, has multiples formas de resolverlo, la primera sería despejar POS.
<br>
pero utilizamos esta página  https://keisan.casio.com/exec/system/14013211335541

Al correrlo obtenemos que  POS = 755 , o sea **las BAJA+2  de 202101  son  755**

¿Qué nos dicen estas 755 BAJA+2 de enero del 2021? ¿Explican algo?
<br>
Las BAJA+2 que hay en noviembre del 2020 son  648 
<br>
755 > 648
<br>
Con lo cual es de esperar mayores ganancias de los modelos predictivos en 202101  que en 202001,  va a irnos mejor en el Public Leaderboard que en nuestros modelos.